packages/oauth/oauth-provider/src/output/send-web-page.ts

```diff
@@ -1,82 +1,51 @@
 ): Promise<void> {
-  const csp = mergeCsp(options.csp, {
-    'default-src': ["'none'"],
+  // @NOTE the csp string might be quite long. In that case it might be tempting
+  // to set it through the http-equiv <meta> in the HTML. However, some
+  // directives cannot be enforced by browsers when set through the meta tag
+  // (e.g. 'frame-ancestors'). Therefore, it's better to set the CSP through the
+  // HTTP header.
+  const csp = mergeCsp(DEFAULT_CSP, inputCsp, {
     'base-uri': options.base?.origin as undefined | `https://${string}`,
-    'script-src': ["'self'", ...assetsToCsp(options.scripts)],
-    'style-src': ["'self'", ...assetsToCsp(options.styles)],
-    'img-src': ["'self'", 'data:', 'https:'],
-    'connect-src': ["'self'"],
-    'upgrade-insecure-requests': true,
-
-    // Prevents the CSP to be embedded in a page <meta>:
-    'frame-ancestors': ["'none'"],
+    'script-src': options.scripts?.map(assetToCsp),
+    'style-src': options.styles?.map(assetToCsp),
   })
 
-  // @NOTE the csp string might become too long. However, since we need to
-  // specify the "frame-ancestors" directive, we can't use a meta tag. For that
-  // reason, we won't try to avoid too long headers and let the proxy throw
-  // in case of a too long header.
-  res.setHeader('Content-Security-Policy', buildCsp(csp))
-
-  // @TODO: make these headers configurable (?)
-  res.setHeader('Permissions-Policy', 'otp-credentials=*, document-domain=()')
-  res.setHeader('Cross-Origin-Embedder-Policy', 'credentialless')
-  res.setHeader('Cross-Origin-Resource-Policy', 'same-origin')
-  res.setHeader('Cross-Origin-Opener-Policy', 'same-origin')
-  res.setHeader('Referrer-Policy', 'same-origin')
-  res.setHeader('X-Frame-Options', 'DENY')
-  res.setHeader('X-Content-Type-Options', 'nosniff')
-  res.setHeader('X-XSS-Protection', '0')
-  res.setHeader('Strict-Transport-Security', 'max-age=63072000')
-
-  const html = buildDocument(options)
-
-  return writeHtml(res, html.toString(), options)
+  const html = buildDocument(options).toString()
+  return writeHtml(res, html, { ...options, csp })
 }
 
-export function* assetsToCsp(
-  assets?: Iterable<Html | AssetRef>,
-): Generator<CspValue> {
-  if (assets) {
-    for (const asset of assets) {
-      yield assetToCsp(asset)
-    }
-  }
-}
-
-export function assetToCsp(asset: Html | AssetRef): CspValue {
+function assetToCsp(asset: Html | AssetRef): CspValue {
   if (asset instanceof Html) {
+    // Inline assets are "allowed" by their hash
     const hash = createHash('sha256').update(asset.toString()).digest('base64')
     return `'sha256-${hash}'`
   } else {
-    return `'sha256-${asset.sha256}'`
+    // External assets are referenced by their full url
+    if (asset.url.startsWith('https:')) return asset.url as `https:${string}`
+    if (asset.url.startsWith('http:')) return asset.url as `http:${string}`
+
+    // Internal assets are served from the same origin
+    return `'self'`
```

Comment:
Bug fix. Specifying the `sha256-` hash of a script that is loaded through `<script src="url"></script>` makes no sense.